rfd/0182-multi-port-tcp-app-access.md

```diff
@@ -0,0 +1,496 @@
+---
+authors: Rafał Cieślak (rafal.cieslak@goteleport.com)
+state: draft
+---
+
+# RFD 182 - Multi-port TCP app access
+
+## Required Approvers
+
+* Engineering: @zmb3 && @nklaassen && (@espadolini || @smallinsky)
+
+## What
+
+Multi-port TCP app access enables cluster administrators to set up TCP apps that expose more than
+just one port to the cluster.
+
+## Why
+
+In the past, many customers expressed interest in specifying a port range for an app
+([#5669](https://github.com/gravitational/teleport/issues/5669), [#10412](https://github.com/gravitational/teleport/issues/10412)).
+Two customers asked about multi-port support immediately after seeing a demo of VNet. Some customers
+have environments where a single logical app is spread across several ports.
+
+Before VNet ([RFD 163](https://github.com/gravitational/teleport/blob/master/rfd/0163-vnet.md)),
+introducing multi-port support was limited by the fact that the end user would need to manually
+start and manage a local proxy per each port. With VNet, this restriction is gone. We can extend
+VNet and TCP app access in such a way that the end user can connect to
+`tcp-app.teleport.example.com:1234` and the connection will be forwarded to the same port number on
+the hostname specified in the app spec.
+
+## Details
+
+### UX
+
+#### Setup
+
+There's an app called waldo which accepts connections on multiple ports. Alice wants to make the app
+available in the teleport.example.com cluster. She creates a new app and includes a new field called
+`ports` in the app definition. She leaves the port number out of the `uri` field.
+
+```yaml
+apps:
+- name: "waldo"
+  uri: "tcp://localhost"
+  ports: [4080-4090, 5095]
+  labels:
+    env: prod
+```
```

Comment:
One more aspect I think we should address in the RFD is whether we should support port mapping, as opposed to the current design where the port VNet will "listen" on must match the "target" port we dial the app with

you could imagine that an app server has an app running locally listening on localhost:8080 and localhost:8443 that they want to expose via vnet on ports 80 and 443. It could be configured something like
```yaml
- name: "waldo"
  uri: "tcp://localhost"
  ports:
  - listen: 80
    target: 8080
  - listen: 443
    target: 8443
  - listen: 5000-5010 # maybe support ranges too
    target: 6000-6010
  - listen: 8081 # if either listen or target is omitted, you could default to matching the one that's set
```

There's slightly more to it but I could see it being a usecase people will want. It would be necessary if they want to expose multiple `localhost` apps on the same ports, which we coincidentally support for single-port apps today by mapping all "listen" ports to the single target port

The slight quirk though is the `listen` address would only be relevant to `vnet`, not `tsh proxy app` - or maybe we could make that the default listen port for `tsh proxy app`? likely to conflict between multiple apps though

Reply:
It's a good idea, but I think I'll leave it for later. As Sasha said in a response to my status email on August 8th, the design partner asks for a pretty narrow scope and there's enough work here already without introducing port mapping.